src/aotcompile.cpp

```diff
@@ -1645,6 +1650,47 @@ void jl_dump_native_impl(void *native_code,
         has_veccall = !!dataM.getModuleFlag("julia.mv.veccall");
     });
 
+#if defined(_OS_LINUX_)
+    // This flag tracks if the lock was actually acquired to avoid freeing the semaphore
+    // if it was never opened.
+    bool lock_acquired = false;
+
+    auto cleanup = [&](sem_t* s) {
+        if (s != SEM_FAILED) {
+            if (lock_acquired) {
+                // Only post if the lock was successfully acquired.
+                sem_post(s);
+            }
+            sem_close(s);
+        }
+    };
+
+    sem_t *par_precomp_sem = SEM_FAILED;
+    const char *sem_name = getenv("JC_PAR_PRECOMP_SEM");
```

Comment:
Can we pick a more verbose name for this environment variable? How about `JL_AOT_PRECOMPILE_SEMAPHORE`?